core/solver/multigrid.cpp

```diff
@@ -159,346 +160,342 @@ void handle_list(
-            kcycle_state.allocate_memory<VT>(level, cycle, current_nrows,
-                                             next_nrows);
-        }
-    }
 
+namespace detail {
+
+
+/**
+ * MultigridState is used to store the necessary cache and run the operation of
+ * all levels.
+ *
+ * @note it should only be used internally
+ */
+struct MultigridState {
+    MultigridState() : nrhs{0} {}
+
+    /**
+     * Generate the cache for later usage.
+     *
+     * @param system_matrix_in  the system matrix
+     * @param multigrid_in  the multigrid information
+     * @param nrhs_in  the number of right hand side
+     */
+    void generate(const LinOp* system_matrix_in, const Multigrid* multigrid_in,
+                  const size_type nrhs_in);
+
+    /**
+     * allocate_memory is a helper function to allocate the memory of one level
+     *
+     * @tparam ValueType  the value type of memory
+     *
+     * @param level  the current level index
+     * @param cycle  the multigrid cycle
+     * @param current_nrows  the number of rows of current fine matrix
+     * @param next_nrows  the number of rows of next coarse matrix
+     */
+    template <typename ValueType>
+    void allocate_memory(int level, multigrid::cycle cycle,
+                         size_type current_nrows, size_type next_nrows);
+
+    /**
+     * run the cycle of the level
+     *
+     * @param cycle  the multigrid cycle
+     * @param level  the current level index
+     * @param matrix  the system matrix of current level
+     * @param b  the right hand side
+     * @param x  the input vectors
+     * @param info  the info of cycle (See cycle_info)
+     */
+    void run_mg_cycle(multigrid::cycle cycle, size_type level,
+                      const std::shared_ptr<const LinOp>& matrix,
+                      const LinOp* b, LinOp* x,
+                      cycle_info info = cycle_info::first_of_cycle |
+                                        cycle_info::end_of_cycle);
+
+    /**
+     * @copydoc run_cycle
+     *
+     * @tparam ValueType  the value type
+     *
+     * @note it is the version with known ValueType
+     */
+    template <typename ValueType>
     void run_cycle(multigrid::cycle cycle, size_type level,
                    const std::shared_ptr<const LinOp>& matrix, const LinOp* b,
-                   LinOp* x, bool x_is_zero = false, bool is_first = true,
-                   bool is_end = true)
-    {
-        if (level == multigrid->get_mg_level_list().size()) {
-            multigrid->get_coarsest_solver()->apply(b, x);
-            return;
-        }
-        auto mg_level = multigrid->get_mg_level_list().at(level);
+                   LinOp* x, cycle_info info);
+
+    // current level's nrows x nrhs
+    std::vector<std::shared_ptr<LinOp>> r_list;
+    // next level's nrows x nrhs
+    std::vector<std::shared_ptr<LinOp>> g_list;
+    std::vector<std::shared_ptr<LinOp>> e_list;
+    // constant 1 x 1
```

Comment:
It seems like the comments no longer match the members? Was this left-over from the recursive storage attempt?

Reply:
the comments is described in the current level fashion, so it also fits in the recursive thought.
level 0 -> r_list[0] has current level's #rows and g_list[0], e_list[0] has next level's #rows